File: internal/utils/blend.go
package utils

import (
	"fmt"
	"image"
	"image/color"
	"math"
)

// opaqueAlpha is the weight at or above which a blend is indistinguishable from a straight write: the result is
// rounded to 8 bits, so anything this close to 1 lands on the source byte anyway. Naming it keeps the threshold in the
// two places that test it from drifting apart.
const opaqueAlpha = 0.999

// ParamIntensity is the map key used by operations to carry the per-run blend amount to Model.Run, decoupled from the
// operation Id so the registry reuses a single session across all intensities.
const ParamIntensity = "intensity"

// IntensityFromParams reads the per-run blend amount from a params map; it defaults to 1.0 (full model output) when the
// key is absent or has the wrong type.
func IntensityFromParams(params map[string]any) float32 {
	if v, ok := params[ParamIntensity].(float32); ok {
		return v
	}
	return 1.0
}

// IntensityCacheKey is the stable per-run signature folded into the image cache key so that runs with the same Id but a
// different blend amount do not collide.
func IntensityCacheKey(intensity float32) string {
	return fmt.Sprintf("i=%.3g", intensity)
}

// BlendWithIntensity blends the original image with the model output based on intensity.
//
//   - intensity = 1.0: full model output
//   - intensity = 0.0: original image
//   - intensity = -1.0: opposite of model adjustment (inverse effect)
func BlendWithIntensity(original, modelOutput image.Image, intensity float32) image.Image {
	bounds := original.Bounds()
	width := bounds.Dx()
	height := bounds.Dy()

	result := image.NewNRGBA(image.Rect(0, 0, width, height))

	// Fast path: when both inputs are concrete RGBA-family buffers, blend via direct Pix indexing. Sample16 reproduces
	// the exact 16-bit values At().RGBA() would return, so the output is bit-identical to the generic path while
	// avoiding per-pixel interface dispatch on two sources.
	//
	// No origin offset appears below: RgbPixBuffer hands back each image's own Pix, which already starts at that
	// image's Bounds().Min, so 0-based indexing is correct whatever the bounds are.
	oPix, oStride, oFast := RgbPixBuffer(original)
	mPix, mStride, mFast := RgbPixBuffer(modelOutput)
	_, oIsNRGBA := original.(*image.NRGBA)
	_, mIsNRGBA := modelOutput.(*image.NRGBA)

	// Both loops are sized from original but index both buffers, so a modelOutput smaller than the original would run
	// off the end of mPix. The generic path reads such pixels as transparent black rather than panicking, so leave the
	// mismatch to it.
	modelBounds := modelOutput.Bounds()
	modelCovers := modelBounds.Dx() >= width && modelBounds.Dy() >= height

	if oFast && mFast && modelCovers {
		for y := range height {
			oRow := y * oStride
			mRow := y * mStride
			dst := y * result.Stride
			for x := range width {
				origR, origG, origB, origA := Sample16(oPix, oRow+x*4, oIsNRGBA)
				modelR, modelG, modelB, _ := Sample16(mPix, mRow+x*4, mIsNRGBA)

				oR := float32(origR) / 257.0
				oG := float32(origG) / 257.0
				oB := float32(origB) / 257.0

				mR := float32(modelR) / 257.0
				mG := float32(modelG) / 257.0
				mB := float32(modelB) / 257.0

				result.Pix[dst] = uint8(Clamp255(oR + intensity*(mR-oR)))
				result.Pix[dst+1] = uint8(Clamp255(oG + intensity*(mG-oG)))
				result.Pix[dst+2] = uint8(Clamp255(oB + intensity*(mB-oB)))
				result.Pix[dst+3] = uint8(origA / 257)
				dst += 4
			}
		}
		return result
	}

	// At() is absolute, so unlike the Pix path above this one does need each image's origin added back.
	origMin, modelMin := bounds.Min, modelBounds.Min

	for y := range height {
		for x := range width {
			origR, origG, origB, origA := original.At(origMin.X+x, origMin.Y+y).RGBA()
			modelR, modelG, modelB, _ := modelOutput.At(modelMin.X+x, modelMin.Y+y).RGBA()

			// Convert to float32 in range [0, 255]
			oR := float32(origR) / 257.0
			oG := float32(origG) / 257.0
			oB := float32(origB) / 257.0

			mR := float32(modelR) / 257.0
			mG := float32(modelG) / 257.0
			mB := float32(modelB) / 257.0

			// Extrapolates past the endpoints for a negative intensity, which is what inverts the effect.
			r := oR + intensity*(mR-oR)
			g := oG + intensity*(mG-oG)
			b := oB + intensity*(mB-oB)

			result.Set(x, y, color.NRGBA{
				R: uint8(Clamp255(r)),
				G: uint8(Clamp255(g)),
				B: uint8(Clamp255(b)),
				A: uint8(origA / 257),
			})
		}
	}

	return result
}

// blendTileWithOverlap blends a tile into a destination image, ramping the edges it shares with tiles already written.
//
// overlapX and overlapY are how many pixels this tile actually overlaps its left and top neighbours; zero means there
// is no neighbour on that side and that edge is written straight. They are passed in rather than assumed because the
// grid shifts a final tile back to sit flush with the image instead of shrinking it, so the last column and row
// overlap by much more than the configured overlap - see TileGrid.OverlapsX.
//
// The ramp is a raised cosine rather than a straight line. Both are continuous, but a linear ramp's slope jumps at
// each end of the band, and on smooth gradients that reads as a visible edge where the blend starts and stops. The
// same curve is used by the diffusion upscaler's own accumulator, for the same reason.
//
// dst is always an *image.RGBA. When src is a concrete RGBA-family type, it is blended via direct Pix indexing instead
// of the image.Image interface, and the fully opaque interior - past both ramps, where the weight is exactly 1 - is
// copied a row at a time. This is output-identical to the per-pixel path but avoids interface dispatch and the float
// blend across what is the overwhelming majority of every tile.
func blendTileWithOverlap(dst *image.RGBA, src image.Image, x, y, overlapX, overlapY int) {
	srcBounds := src.Bounds()

	dstBounds := dst.Bounds()

	// Pre-calculate actual rendering bounds to avoid repeated checks
	maxX := min(dstBounds.Dx()-x, srcBounds.Dx())
	maxY := min(dstBounds.Dy()-y, srcBounds.Dy())
	if maxX <= 0 || maxY <= 0 {
		return
	}

	// A ramp wider than half the tile would have its two ends overlap each other; clamping matches what the diffusion
	// upscaler's edgeWeights does and keeps each ramp monotonic.
	rampX := min(overlapX, maxX/2)
	rampY := min(overlapY, maxY/2)

	srcPix, srcStride, srcFast := RgbPixBuffer(src)
	_, srcIsNRGBA := src.(*image.NRGBA)

	// The bulk path reproduces the per-pixel one only for a premultiplied buffer: an NRGBA tile with a non-opaque
	// pixel needs the premultiply below, which a straight copy would skip.
	bulk := srcFast && !srcIsNRGBA

	for dy := range maxY {
		dstY := y + dy
		srcY := srcBounds.Min.Y + dy

		topAlpha := rampWeight(dy, rampY)

		// The horizontal ramp only affects the leftmost rampX columns; everything to their right shares this row's
		// weight - 1, or topAlpha while the row is still inside the top band.
		dstRow := dst.PixOffset(x, dstY)

		if bulk && topAlpha >= opaqueAlpha {
			// Past both ramps the tile simply replaces what is underneath, so the run is a copy. The alpha byte is
			// forced afterwards rather than taken from the source, matching what the per-pixel path writes.
			from := dstRow + rampX*4
			to := dstRow + maxX*4
			si := dy*srcStride + rampX*4

			copy(dst.Pix[from:to], srcPix[si:si+(maxX-rampX)*4])

			for p := from + 3; p < to; p += 4 {
				dst.Pix[p] = 255
			}

			if rampX == 0 {
				continue
			}

			blendRun(dst, srcPix, srcStride, src, srcBounds, dstRow, dy, srcY, 0, rampX, rampX, topAlpha,
				srcFast, srcIsNRGBA)

			continue
		}

		blendRun(dst, srcPix, srcStride, src, srcBounds, dstRow, dy, srcY, 0, maxX, rampX, topAlpha,
			srcFast, srcIsNRGBA)
	}
}

// blendRun blends columns [from, to) of one row. It is the per-pixel path, shared by the rows the bulk copy cannot
// take and by the left ramp of the rows it can.
func blendRun(
	dst *image.RGBA,
	srcPix []uint8,
	srcStride int,
	src image.Image,
	srcBounds image.Rectangle,
	dstRow, dy, srcY, from, to, rampX int,
	topAlpha float64,
	srcFast, srcIsNRGBA bool,
) {
	for dx := from; dx < to; dx++ {
		// The two ramps multiply where they meet, so a corner covered by both is weighted by each.
		alpha := topAlpha
		if dx < rampX {
			alpha *= rampWeight(dx, rampX)
		}

		var sr, sg, sb uint32
		if srcFast {
			// srcPix already starts at src.Bounds().Min, so this indexes from the tile's own origin. Adding
			// srcBounds.Min back - as the At() fallback below correctly must - would apply it twice.
			si := dy*srcStride + dx*4
			sr, sg, sb = uint32(srcPix[si]), uint32(srcPix[si+1]), uint32(srcPix[si+2])
			// RGBA buffers are already premultiplied 8-bit; only NRGBA with non-opaque alpha needs
			// the premultiply that RGBA() would apply (matches the >>8 of the 16-bit result).
			if srcIsNRGBA {
				if a := uint32(srcPix[si+3]); a != 0xff {
					sr = ((sr * 257) * a / 0xff) >> 8
					sg = ((sg * 257) * a / 0xff) >> 8
					sb = ((sb * 257) * a / 0xff) >> 8
				}
			}
		} else {
			r, g, b, _ := src.At(srcBounds.Min.X+dx, srcY).RGBA()
			// Match the 8-bit values the slow path historically used (RGBA() >> 8).
			sr, sg, sb = r>>8, g>>8, b>>8
		}

		di := dstRow + dx*4
		if alpha >= opaqueAlpha {
			dst.Pix[di] = uint8(sr)
			dst.Pix[di+1] = uint8(sg)
			dst.Pix[di+2] = uint8(sb)
			dst.Pix[di+3] = 255
		} else {
			invAlpha := 1.0 - alpha
			dst.Pix[di] = uint8(float64(sr)*alpha + float64(dst.Pix[di])*invAlpha)
			dst.Pix[di+1] = uint8(float64(sg)*alpha + float64(dst.Pix[di+1])*invAlpha)
			dst.Pix[di+2] = uint8(float64(sb)*alpha + float64(dst.Pix[di+2])*invAlpha)
			dst.Pix[di+3] = 255
		}
	}
}

// rampWeight is the incoming tile's weight at offset i into a ramp of the given width, and 1 once past it (a zero
// width being "no neighbour on this side").
//
// The half-pixel offset keeps the first weight strictly positive: at i = 0 a bare cosine would be exactly 0, throwing
// away the incoming tile's outermost column entirely rather than mixing it.
func rampWeight(i, width int) float64 {
	if width <= 0 || i >= width {
		return 1
	}

	return 0.5 - 0.5*math.Cos(math.Pi*(float64(i)+0.5)/float64(width))
}

File: crates/imaging/src/blend.rs
//! Writing a finished tile into the picture without leaving a seam where it meets the last one.
//!
//! **One of this crate's two ways of combining tiles**, and the one for a model whose output is deterministic: each
//! tile is written *over* what is already in the buffer, at the caller's channel depth. A model drawing its own noise
//! per tile uses [`canvas`](crate::canvas) instead.

use image::{ImageBuffer, Rgb};

use crate::tensor::Channel;

// Derived from the channel's depth rather than written as one figure, because what "indistinguishable" means is half a
// level, and a level is 1/255 of the range at eight bits and 1/65535 at sixteen. Above this weight the blend moves the
// incoming value by less than half a level, so rounding lands it back on that value exactly. A fixed 0.999 held at
// eight bits and not at sixteen, where the 0.001 it waved through is 65 levels. Naming it keeps the threshold and the
// fast path that tests it from drifting apart, and it is what makes an unblended edge a byte-for-byte copy rather than
// a round trip through a float.
/// The weight above which a blend at `T`'s depth is indistinguishable from a straight write.
fn opaque<T: Channel>() -> f32
where
    Rgb<T>: image::Pixel<Subpixel = T>,
{
    1.0 - 0.5 / T::MAX_VALUE
}

/// The incoming tile's weight at `offset` into a ramp `width` pixels wide, and 1 once past it — a zero width being
/// "no neighbour on this side". A raised cosine, strictly positive from the first offset.
#[inline]
pub fn ramp_weight(offset: u32, width: u32) -> f32 {
    // Usable without `blend_tile`, because `canvas` feathers the diffusion upscaler's tiles into a weighted float
    // accumulator with this same curve. The two agreeing on it is load-bearing — a tuning change applied to only one
    // of them shows up as a seam.
    //
    // A raised cosine rather than a straight line. Both are continuous, but a linear ramp's slope jumps at each end of
    // the band, and on a smooth gradient that reads as a visible edge exactly where the blend starts and stops.
    if width == 0 || offset >= width {
        return 1.0;
    }

    // The half-pixel offset is what keeps the first weight strictly positive: at offset 0 a bare cosine is exactly 0,
    // which throws the incoming tile's outermost column away entirely rather than mixing it in.
    let position = (f64::from(offset) + 0.5) / f64::from(width);

    (0.5 - 0.5 * (std::f64::consts::PI * position).cos()) as f32
}

/// Writes `tile` into `dest` at `(x, y)`, ramping each edge that meets an already-written neighbour.
///
/// `overlap_x` and `overlap_y` are how many pixels this tile **actually** overlaps its left and top neighbours — not
/// the configured overlap; see [`Layout::overlaps_x`](crate::grid::Layout::overlaps_x) — and zero means there is no
/// neighbour on that side and that edge is written straight.
///
/// Where a pixel lies in a band on both axes the two weights multiply, so a corner is weighted by each.
///
/// A tile that runs past `dest` is written only as far as `dest` goes.
pub fn blend_tile<T: Channel>(
    dest: &mut ImageBuffer<Rgb<T>, Vec<T>>,
    tile: &ImageBuffer<Rgb<T>, Vec<T>>,
    x: u32,
    y: u32,
    overlap_x: u32,
    overlap_y: u32,
) where
    Rgb<T>: image::Pixel<Subpixel = T>,
{
    let columns = dest.width().saturating_sub(x).min(tile.width());
    let rows = dest.height().saturating_sub(y).min(tile.height());

    if columns == 0 || rows == 0 {
        return;
    }

    // A ramp wider than half the tile would have its two ends overlap each other and stop being monotonic, which
    // here is a visible band: this writes over what is underneath, so a weight that dips back down is the picture
    // dipping with it. `canvas::axis_weights` does **not** clamp, and says why.
    let ramp_x = overlap_x.min(columns / 2);
    let ramp_y = overlap_y.min(rows / 2);

    let opaque = opaque::<T>();

    // Tabulated once per tile rather than evaluated per pixel: the horizontal weight depends only on the column, so
    // the loop below would otherwise recompute the same cosines on every one of `rows` rows. The vertical weight is
    // already a per-row value, so it stays a direct call.
    let horizontal: Vec<f32> = (0..ramp_x).map(|offset| ramp_weight(offset, ramp_x)).collect();

    // The strides the row copy below indexes with. Read before `dest` is borrowed mutably.
    let (dest_width, tile_width) = (dest.width() as usize, tile.width() as usize);

    for row in 0..rows {
        let vertical = ramp_weight(row, ramp_y);

        // Past the vertical ramp every column at or beyond `ramp_x` carries weight 1: the tile simply replaces what
        // is underneath. That is the large majority of every tile — at the default geometry a 4x pass decodes to
        // 1024x1024, of which the ramp band is two strips 16 pixels wide — and it is a copy, so it is done as one.
        // Per pixel it would be two bounds checks and an index recomputation each, four hundred tiles to a pass.
        if vertical > opaque && ramp_x < columns {
            let span = ((columns - ramp_x) * 3) as usize;
            let from = (row as usize * tile_width + ramp_x as usize) * 3;
            let into = ((y + row) as usize * dest_width + (x + ramp_x) as usize) * 3;

            // Through the buffers' own sample slices: `ImageBuffer`'s `Index` is by coordinate, so the raw samples
            // have to be reached for explicitly. Both are `Rgb<T>` with a stride of `width * 3`, so the spans line up.
            let written: &mut [T] = dest;
            written[into..into + span].copy_from_slice(&(**tile)[from..from + span]);
        }

        // The ramp band, and — on a row still inside the vertical ramp — the whole width, since there the tile is
        // mixed with what is underneath at every column.
        let blended_to = if vertical > opaque { ramp_x } else { columns };

        for column in 0..blended_to {
            let weight = match horizontal.get(column as usize) {
                Some(across) => vertical * across,
                None => vertical,
            };

            let source = *tile.get_pixel(column, row);

            let blended = if weight > opaque {
                // The tile replaces what is underneath, so the value is carried across unchanged rather than round
                // tripped through a float.
                source
            } else {
                let under = *dest.get_pixel(x + column, y + row);
                let mut mixed = source;

                for channel in 0..3 {
                    let over = source.0[channel].to_unit();
                    mixed.0[channel] = T::from_unit(under.0[channel].to_unit().mul_add(1.0 - weight, over * weight));
                }

                mixed
            };

            dest.put_pixel(x + column, y + row, blended);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A uniform buffer at a `[0, 1]` level, so one scenario can be stated once and run at both depths.
    fn flat<T: Channel>(width: u32, height: u32, level: f32) -> ImageBuffer<Rgb<T>, Vec<T>>
    where
        Rgb<T>: image::Pixel<Subpixel = T>,
    {
        let value = T::from_unit(level);
        ImageBuffer::from_pixel(width, height, Rgb([value, value, value]))
    }

    /// The red channel of one destination row, as `[0, 1]` levels — the units every assertion below is written in,
    /// so the same tolerances hold at eight bits and at sixteen.
    fn row<T: Channel>(dest: &ImageBuffer<Rgb<T>, Vec<T>>, y: u32) -> Vec<f32>
    where
        Rgb<T>: image::Pixel<Subpixel = T>,
    {
        (0..dest.width()).map(|x| dest.get_pixel(x, y).0[0].to_unit()).collect()
    }

    /// The largest jump between adjacent columns of `levels` over `range`.
    fn largest_step(levels: &[f32], range: std::ops::Range<usize>) -> f32 {
        levels[range].windows(2).map(|pair| (pair[1] - pair[0]).abs()).fold(0.0, f32::max)
    }

    /// A flat tile written over a differently-coloured background across a seam 120 pixels wide — the width the last
    /// column of a real grid actually shares, rather than the 16 it was configured with.
    fn a_wide_seam_leaves_no_step<T: Channel>()
    where
        Rgb<T>: image::Pixel<Subpixel = T>,
    {
        let mut dest = flat::<T>(400, 4, 40.0 / 255.0);
        let tile = flat::<T>(240, 4, 200.0 / 255.0);

        blend_tile(&mut dest, &tile, 100, 0, 120, 0);
        let levels = row(&dest, 0);

        // Across the whole seam — the column before the tile starts, through the ramp, to the first column past it.
        let step = largest_step(&levels, 99..222);
        assert!(step < 0.02, "a 120-pixel seam left a {step} step between adjacent columns");

        // The ramp actually spans the seam rather than ending early: it starts near the background and reaches the
        // tile's own level by the far side of the band.
        assert!((levels[100] - 40.0 / 255.0).abs() < 0.01, "the seam did not start at the background");
        assert!((levels[220] - 200.0 / 255.0).abs() < 0.01, "the seam did not reach the tile by its far side");

        // And the regression this figure exists to prevent: ramping the same seam over the *configured* 16 pixels
        // leaves a step several times larger, which is the hard edge down the last column of the picture.
        let mut narrow = flat::<T>(400, 4, 40.0 / 255.0);
        blend_tile(&mut narrow, &tile, 100, 0, 16, 0);

        let narrow_step = largest_step(&row(&narrow, 0), 99..222);
        assert!(narrow_step > step * 3.0, "ramping a 120-pixel seam over 16 pixels left no larger step");
    }

    /// An edge with no already-written neighbour is written at full strength, byte for byte.
    fn an_edge_with_no_neighbour_is_written_unchanged<T: Channel + std::fmt::Debug>()
    where
        Rgb<T>: image::Pixel<Subpixel = T>,
    {
        let mut dest = flat::<T>(64, 64, 0.0);
        let mut tile = flat::<T>(32, 32, 0.75);
        tile.put_pixel(0, 0, Rgb([T::from_unit(0.25), T::from_unit(0.5), T::from_unit(1.0)]));

        blend_tile(&mut dest, &tile, 0, 0, 0, 0);

        for y in 0..32 {
            for x in 0..32 {
                assert_eq!(dest.get_pixel(x, y), tile.get_pixel(x, y), "the first tile was altered at ({x}, {y})");
            }
        }

        // Nothing outside the tile was touched to weight it.
        assert_eq!(*dest.get_pixel(32, 0), Rgb([T::from_unit(0.0); 3]));
    }

    /// A pixel inside a band on both axes is weighted by each of them.
    fn a_corner_is_weighted_by_both_axes<T: Channel>()
    where
        Rgb<T>: image::Pixel<Subpixel = T>,
    {
        let mut dest = flat::<T>(200, 200, 0.0);
        let tile = flat::<T>(100, 100, 1.0);

        blend_tile(&mut dest, &tile, 50, 50, 20, 20);

        // Over a black background the level a pixel reaches is its weight. Read ten pixels into both bands, where
        // the two weights are far enough from their endpoints that either depth can tell them apart.
        let corner = dest.get_pixel(60, 60).0[0].to_unit();
        let in_column_only = dest.get_pixel(60, 90).0[0].to_unit();
        let in_row_only = dest.get_pixel(90, 60).0[0].to_unit();

        let across = ramp_weight(10, 20);
        assert!(
            (corner - across * across).abs() < 0.01,
            "the corner is at {corner}, not the product of both bands"
        );
        assert!((in_column_only - across).abs() < 0.01, "a pixel in one band only carries more than that band");
        assert!(corner < in_column_only - 0.1, "the corner was not weighted by the vertical band as well");
        assert!(corner < in_row_only - 0.1, "the corner was not weighted by the horizontal band as well");

        // And the outermost corner, where both weights are at their smallest, is the product there too.
        let outermost = dest.get_pixel(50, 50).0[0].to_unit();
        let expected = ramp_weight(0, 20) * ramp_weight(0, 20);
        assert!((outermost - expected).abs() < 0.01, "the outermost corner is at {outermost}, not {expected}");

        // Past both bands the tile is at full strength.
        assert!((dest.get_pixel(90, 90).0[0].to_unit() - 1.0).abs() < 0.01);
    }

    /// A band wider than half the tile is clamped, which is what keeps each ramp monotonic.
    fn a_band_wider_than_half_the_tile_is_clamped<T: Channel>()
    where
        Rgb<T>: image::Pixel<Subpixel = T>,
    {
        let mut dest = flat::<T>(100, 100, 0.0);
        let tile = flat::<T>(40, 40, 1.0);

        blend_tile(&mut dest, &tile, 0, 0, 400, 400);
        let levels = row(&dest, 20);

        for pair in levels[..40].windows(2) {
            assert!(pair[1] >= pair[0], "an over-wide band stopped being monotonic at {pair:?}");
        }

        // Clamped to half the tile, so the far half is at full strength rather than still ramping.
        assert!(
            (levels[39] - 1.0).abs() < 0.01,
            "the clamped band did not reach full strength by the tile's end"
        );
    }

    #[test]
    fn every_blend_scenario_holds_at_eight_bits() {
        a_wide_seam_leaves_no_step::<u8>();
        an_edge_with_no_neighbour_is_written_unchanged::<u8>();
        a_corner_is_weighted_by_both_axes::<u8>();
        a_band_wider_than_half_the_tile_is_clamped::<u8>();
    }

    #[test]
    fn every_blend_scenario_holds_at_sixteen_bits() {
        a_wide_seam_leaves_no_step::<u16>();
        an_edge_with_no_neighbour_is_written_unchanged::<u16>();
        a_corner_is_weighted_by_both_axes::<u16>();
        a_band_wider_than_half_the_tile_is_clamped::<u16>();
    }

    #[test]
    fn a_weight_just_short_of_one_is_blended_at_sixteen_bits_and_copied_at_eight() {
        // The last column of a 40-pixel band weighs about 0.9996: close enough to 1 that eight bits cannot tell, and
        // 25 levels short of it at sixteen — which a fixed 0.999 cutoff used to copy straight across.
        let weight = ramp_weight(39, 40);
        assert!(weight > 0.999 && weight < opaque::<u16>(), "the fixture no longer sits between the two cutoffs");

        let mut deep = flat::<u16>(100, 1, 0.0);
        blend_tile(&mut deep, &flat::<u16>(100, 1, 1.0), 0, 0, 40, 0);
        assert_eq!(deep.get_pixel(39, 0).0[0], u16::from_unit(weight), "sixteen bits copied a blended column");

        let mut shallow = flat::<u8>(100, 1, 0.0);
        blend_tile(&mut shallow, &flat::<u8>(100, 1, 1.0), 0, 0, 40, 0);
        assert_eq!(shallow.get_pixel(39, 0).0[0], u8::MAX, "eight bits changed a column it cannot resolve");
    }

    #[test]
    fn a_tile_that_runs_past_the_destination_is_written_only_as_far_as_it_goes() {
        // Which is what crops the mirrored padding off the one tile that can carry any.
        let mut dest = flat::<u8>(10, 10, 0.0);
        let tile = flat::<u8>(16, 16, 1.0);

        blend_tile(&mut dest, &tile, 4, 4, 0, 0);

        assert_eq!(dest.dimensions(), (10, 10));
        assert_eq!(*dest.get_pixel(9, 9), Rgb([255, 255, 255]));
        assert_eq!(*dest.get_pixel(3, 3), Rgb([0, 0, 0]));
    }

    #[test]
    fn the_curve_rises_from_above_zero_to_one_across_the_band() {
        for width in [1, 2, 16, 120, 256] {
            let weights: Vec<f32> = (0..width).map(|offset| ramp_weight(offset, width)).collect();

            assert!(weights[0] > 0.0, "the first weight of a {width}-pixel band throws the outermost column away");

            for pair in weights.windows(2) {
                assert!(pair[1] > pair[0], "a {width}-pixel band is not monotonic at {pair:?}");
            }

            let last = weights[weights.len() - 1];
            assert!(last < 1.0, "a {width}-pixel band reaches full strength inside the band");
            assert!(last > 1.0 - 2.0 / width as f32, "a {width}-pixel band ends at {last}, short of full strength");
        }
    }

    #[test]
    fn past_the_band_the_incoming_tile_is_at_full_strength() {
        assert_eq!(ramp_weight(16, 16), 1.0);
        assert_eq!(ramp_weight(17, 16), 1.0);
        assert_eq!(ramp_weight(u32::MAX, 16), 1.0);
    }

    #[test]
    fn a_band_of_no_width_is_full_strength_everywhere() {
        // Which is what "this edge has no already-written neighbour" is spelt as.
        for offset in [0, 1, 255] {
            assert_eq!(ramp_weight(offset, 0), 1.0, "a zero-width band weighted the tile at {offset}");
        }
    }

    #[test]
    fn the_curve_is_symmetric_about_the_middle_of_the_band() {
        // The property that makes it a *raised cosine* rather than any other rising curve: the weight it gives the
        // incoming tile and the weight the same offset from the far end gives what is underneath sum to 1.
        for offset in 0..120 {
            let rising = ramp_weight(offset, 120);
            let falling = ramp_weight(119 - offset, 120);

            assert!((rising + falling - 1.0).abs() < 1e-6, "the band is not symmetric at {offset}");
        }
    }
}
